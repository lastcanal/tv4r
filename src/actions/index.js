import { extractPosts, filterPosts } from '../helpers'
import reddit from '../helpers/reddit'

import {
  SELECT_SUBREDDIT, INVALIDATE_SUBREDDIT,
  REQUEST_POSTS, RECEIVE_POSTS, RECEIVE_POSTS_ERROR,
  SELECT_POST, NEXT_POST, PREVIOUS_POST,
  MEDIA_FALLBACK
} from '../constants'

export const selectSubreddit = subreddit => ({
  type: SELECT_SUBREDDIT,
  subreddit
})

export const invalidateSubreddit = subreddit => ({
  type: INVALIDATE_SUBREDDIT,
  subreddit
})

export const requestPosts = subreddit => ({
  type: REQUEST_POSTS,
  subreddit
})

export const receivePosts = (subreddit, json) => ({
  type: RECEIVE_POSTS,
  subreddit,
  posts: filterPosts(extractPosts(json)),
  error: null,
  receivedAt: Date.now()
})

export const receivePostsError = (subreddit, error) => ({
  type: RECEIVE_POSTS_ERROR,
  subreddit,
  error
})


export const selectPost = (post, index) => ({
  type: SELECT_POST,
  post,
  index
})

export const nextPost = (posts) => ({
  type: NEXT_POST,
  posts
})

export const previousPost = (posts) => ({
  type: PREVIOUS_POST,
  posts
})

export const mediaFallback = () => ({
  type: MEDIA_FALLBACK
})

const fetchPosts = subreddit => dispatch => {
  dispatch(requestPosts(subreddit))
  return reddit.fetchPosts(subreddit)
    .then(response => response.json())
    .then(response => dispatch(receivePosts(subreddit, response)))
    .catch(error => dispatch(receivePostsError(subreddit, error)))
}

const shouldFetchPosts = (state, subreddit) => {
  const posts = state.postsBySubreddit[subreddit]
  if (!posts) {
    return true
  }
  if (posts.isFetching) {
    return false
  }
  return posts.didInvalidate
}

export const fetchPostsIfNeeded = subreddit => (dispatch, getState) => {
  if (shouldFetchPosts(getState(), subreddit)) {
    return dispatch(fetchPosts(subreddit))
  }
}

