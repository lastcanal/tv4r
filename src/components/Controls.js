import React from 'react'
import PropTypes from 'prop-types'
import { makeStyles, styled } from '@material-ui/core/styles'
import { IconButton, Tooltip } from '@material-ui/core'
import { connect } from 'react-redux'

import SkipNextIcon from '@material-ui/icons/SkipNext'
import SkipPreviousIcon from '@material-ui/icons/SkipPrevious'
import RefreshIcon from '@material-ui/icons/Refresh'
import FullscreenIcon from '@material-ui/icons/Fullscreen'
import FullscreenExitIcon from '@material-ui/icons/FullscreenExit'
import SettingsBrightnessIcon from '@material-ui/icons/SettingsBrightness'
import PlayArrowIcon from '@material-ui/icons/PlayArrow'
import StopIcon from '@material-ui/icons/Stop'
import SyncIcon from '@material-ui/icons/Sync'
import SyncDisabledIcon from '@material-ui/icons/SyncDisabled'
import PhotoSizeSelectActualIcon from '@material-ui/icons/PhotoSizeSelectActual'
import MovieIcon from '@material-ui/icons/Movie'

import {
  nextPost,
  previousPost,
  configToggleFullscreen,
  configToggleAutoAdvance,
  configTogglePlay,
  configToggleThemeMode,
  configToggleShowVideos,
  configToggleShowImages,
  refreshSubreddit,
} from '../actions'

const AutoPlayOnIcon = styled(SyncIcon)({
  transform: 'scale(-1, 1) rotate(135deg)',
})

const AutoPlayOffIcon = styled(SyncDisabledIcon)({
  transform: 'scale(-1, 1) rotate(-90deg)',
})

const useStyles = makeStyles(({ breakpoints }) => ({
  controls: {
    width: '100%',
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    overflow: 'auto',
    marginBottom: 0,
    [breakpoints.up('md')]: {
      justifyContent: 'flex-end',
    },
  },
}))

export const ShowImagesControl = ({ showImages, onClick }) => (
  <Tooltip title="Filter Images">
    <IconButton
      aria-label="Filter Images"
      color="inherit"
      onClick={onClick}
    >
      <PhotoSizeSelectActualIcon
        style={{ opacity: showImages ? 1 : 0.5 }}
      />
    </IconButton>
  </Tooltip>
)

ShowImagesControl.propTypes = {
  showImages: PropTypes.bool,
  onClick: PropTypes.func,
}

export const ShowVideosControl = ({ showVideos, onClick }) => (
  <Tooltip title="Filter Videos">
    <IconButton
      aria-label="Filter Videos"
      color="inherit"
      onClick={onClick}
    >
      <MovieIcon
        style={{ opacity: showVideos ? 1 : 0.5 }}
      />
    </IconButton>
  </Tooltip>
)

ShowVideosControl.propTypes = {
  showVideos: PropTypes.bool,
  onClick: PropTypes.func,
}

export const RefreshControl = ({ onClick }) => (
  <Tooltip title="Refresh Subreddit">
    <IconButton
      aria-label="refresh content"
      color="inherit"
      onClick={onClick}
    >
      <RefreshIcon />
    </IconButton>
  </Tooltip>
)

RefreshControl.propTypes = {
  onClick: PropTypes.func,
}

const Controls = ({
  dispatch,
  posts,
  isFullscreen,
  isAutoAdvance,
  isPlaying,
  themeMode,
  showVideos,
  showImages,
}) => {
  const { controls } = useStyles()

  return (
    <div className={controls}>
      <ShowImagesControl
        showImages={showImages}
        onClick={() => dispatch(configToggleShowImages())}
      />
      <ShowVideosControl
        showVideos={showVideos}
        onClick={() => dispatch(configToggleShowVideos())}
      />
      <RefreshControl onClick={() => dispatch(refreshSubreddit())} />
      <Tooltip
        title={`Switch to ${themeMode === 'dark' ? 'Light' : 'Dark'} theme`}
      >
        <IconButton
          aria-label={`toggle theme`} color="inherit"
          onClick={() => dispatch(configToggleThemeMode())}
        >
          <SettingsBrightnessIcon />
        </IconButton>
      </Tooltip>
      <Tooltip title={`${isFullscreen ? 'Exit' : 'Enter'} Fullscreen`}>
        <IconButton
          aria-label="toggle fullscreen"
          color="inherit"
          onClick={() => dispatch(configToggleFullscreen())}
        >
          {isFullscreen ? <FullscreenExitIcon /> : <FullscreenIcon />}
        </IconButton>
      </Tooltip>
      <Tooltip
        title={`${isAutoAdvance ? 'Stop' : 'Start'} Auto Advancing Videos`}
      >
        <IconButton
          aria-label="toggle auto advance" color="inherit"
          onClick={() => dispatch(configToggleAutoAdvance())}
        >
          {isAutoAdvance ? <AutoPlayOnIcon /> : <AutoPlayOffIcon />}
        </IconButton>
      </Tooltip>
      <Tooltip title="Play Previous">
        <IconButton
          aria-label="play previous"
          color="inherit"
          onClick={() => dispatch(previousPost(posts))}
        >
          <SkipPreviousIcon />
        </IconButton>
      </Tooltip>
      <Tooltip title={`${isAutoAdvance ? 'Start' : 'Stop'} Playing Video`}>
        <IconButton
          aria-label={isPlaying ? 'Stop' : 'Play'} color="inherit"
          onClick={() => dispatch(configTogglePlay())}
        >
          {isPlaying ? <StopIcon /> : <PlayArrowIcon />}
        </IconButton>
      </Tooltip>
      <Tooltip title="Play Next">
        <IconButton
          aria-label="play next"
          color="inherit"
          onClick={() => dispatch(nextPost(posts))}
        >
          <SkipNextIcon />
        </IconButton>
      </Tooltip>
    </div>
  )
}

Controls.propTypes = {
  dispatch: PropTypes.func.isRequired,
  posts: PropTypes.array,
  selectedSubreddit: PropTypes.string,
  isFullscreen: PropTypes.bool,
  isAutoAdvance: PropTypes.bool,
  isPlaying: PropTypes.bool,
  themeMode: PropTypes.string,
  showVideos: PropTypes.bool,
  showImages: PropTypes.bool,
}

const mapStateToProps = state => {
  const { selectedSubreddit, postsBySubreddit, config } = state
  const subreddit = postsBySubreddit[selectedSubreddit]
  const posts = subreddit?.[subreddit?.scope]

  return {
    posts,
    ...config,
  }
}

export default Controls |> connect(mapStateToProps)
