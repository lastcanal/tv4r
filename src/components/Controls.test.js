import SkipNextIcon from '@material-ui/icons/SkipNext'
import SkipPreviousIcon from '@material-ui/icons/SkipPrevious'
import RefreshIcon from '@material-ui/icons/Refresh'

import Controls from './Controls'
import * as actions from '../actions'

describe('Controls', () => {
  const posts = [{ id: 1 }, { id: 2 }]

  it('should render Controls', () => {
    const wrapper = mount(<Controls />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should render handle next post', () => {
    const dispatch = jest.fn(() => {})
    const wrapper = mount(<Controls dispatch={dispatch} posts={posts} />)

    wrapper.find(SkipNextIcon).simulate('click')

    expect(dispatch).toBeCalledWith(actions.nextPost(posts))
  })

  it('should render handle previous post', () => {
    const dispatch = jest.fn(() => {})
    const wrapper = mount(<Controls dispatch={dispatch} posts={posts} />)

    wrapper.find(SkipPreviousIcon).simulate('click')

    expect(dispatch).toBeCalledWith(actions.previousPost(posts))
  })

  it('should render handle refresh', () => {
    const dispatch = jest.fn(() => {})
    const subreddit = 'foo'
    const wrapper = mount(
      <Controls dispatch={dispatch} selectedSubreddit={subreddit} />,
    )

    wrapper.find(RefreshIcon).simulate('click')

    expect(dispatch).toBeCalledWith(actions.invalidateSubreddit(subreddit))

    expect(dispatch).toBeCalledWith(actions.invalidateSubreddit(subreddit))
  })
})
