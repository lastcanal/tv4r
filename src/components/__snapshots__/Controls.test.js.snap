// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls should render Controls 1`] = `ReactWrapper {}`;
